File: backends/Contracts/MemoryCache.Default/Handlers/Set.cs
using D2.Contracts.Handler;
using D2.Contracts.Result;
using Microsoft.Extensions.Caching.Memory;
using S = D2.Contracts.Interfaces.ICommonCacheService;

namespace D2.Contracts.MemoryCache.Default.Handlers;

/// <inheritdoc cref="S.ISetHandler{TValue}"/>
public class Set<TValue> : BaseHandler<
        S.ISetHandler<TValue>, S.SetInput<TValue>, S.SetOutput>,
    S.ISetHandler<TValue>
{
    /// <summary>
    /// Initializes a new instance of the <see cref="Set{TValue}"/> class.
    /// </summary>
    ///
    /// <inheritdoc/>
    public Set(
        IMemoryCache memoryCache,
        IHandlerContext context) : base(context)
    {
        r_memoryCache = memoryCache;
    }

    private readonly IMemoryCache r_memoryCache;

    /// <inheritdoc/>
    protected override ValueTask<D2Result<S.SetOutput?>> ExecuteAsync(
        S.SetInput<TValue> input,
        CancellationToken ct = default)
    {
        if (input.Expiration.HasValue)
            r_memoryCache.Set(input.Key, input.Value, input.Expiration.Value);
        else
            r_memoryCache.Set(input.Key, input.Value);

        return ValueTask.FromResult(D2Result<S.SetOutput?>.Ok(new S.SetOutput(), traceId: TraceId));
    }
}

File: backends/node/shared/implementations/caching/in-memory/default/src/handlers/u/set.ts
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { InMemoryCache } from "@d2/interfaces";
import type { MemoryCacheStore } from "../../memory-cache-store.js";

export class Set<TValue>
  extends BaseHandler<InMemoryCache.SetInput<TValue>, InMemoryCache.SetOutput>
  implements InMemoryCache.ISetHandler<TValue>
{
  private readonly store: MemoryCacheStore;

  constructor(store: MemoryCacheStore, context: IHandlerContext) {
    super(context);
    this.store = store;
  }

  protected async executeAsync(
    input: InMemoryCache.SetInput<TValue>,
  ): Promise<D2Result<InMemoryCache.SetOutput | undefined>> {
    this.store.set(input.key, input.value, input.expirationMs);
    return D2Result.ok({ data: {} });
  }
}
